### PR TITLE
feat: configurable headers

### PR DESCRIPTION
**Pull request by lasiar**
See: #12

### CODE REVIEW

File: testdata/src/initialism/initialism.go [hunk 1]
```diff
@@ -6,69 +6,69 @@ import "net/http"
...
-	h.Get("WWW-Authenticate")
-	h.Get("X-Correlation-ID")
-	h.Get("X-DNS-Prefetch-Control")
-	h.Get("X-Real-IP")
-	h.Get("X-Request-ID")
-	h.Get("X-UA-Compatible")
-	h.Get("X-WebKit-CSP")
-	h.Get("X-XSS")
-	h.Get("X-XSS-Protection")
+	h.Get("A-Im")                              // want `initialism header "A-Im", instead use: "A-IM"`
```
Comment (ccoVeille) [inline testdata/src/initialism/initialism.go:9]: This error message is unclear.

It's the same with non-canonical and the others.

Maybe something like


```suggestion
	h.Get("A-Im")                              // want `"A-Im" header should be "A-IM" (initialism)`
```

Or canonical, I won't add the negation to canonical. Because initialism and canonical are descriptive, when it's written without the expected case.

File: testdata/src/initialism/initialism.go [hunk 2]
```diff
@@ -6,69 +6,69 @@ import "net/http"
 func _() {
```
[the remark this answers is not in the document]
Comment (lasiar) [reply testdata/src/initialism/initialism.go:6]: Don't need it, because this file is located in the testdata folder

File: analyzer.go [hunk 3]
```diff
@@ -19,19 +20,87 @@ const (
...
+	reasonReportInitialism
+)
+
+func (r reasonReport) String() string {
+	switch r {
+	case reasonReportNonCanonical:
+		return "non-canonical"
+
+	case reasonReportExclusion:
+		return "exclusion"
```
Comment (ccoVeille) [inline analyzer.go:61]: This text is displayed I'm the error message, no?

The look at https://github.com/lasiar/canonicalheader/pull/60#discussion_r1835526951

What about this?

```suggestion
		return "excluded"
```

File: constant_string.go [hunk 4]
```diff
@@ -32,14 +32,15 @@ func newConstantKey(info *types.Info, ident *ast.Ident) (constantString, error)
...
 }
 
-func (c constantString) diagnostic(canonicalHeader string) analysis.Diagnostic {
+func (c constantString) diagnostic(canonicalHeader string, r reasonReport) analysis.Diagnostic {
 	return analysis.Diagnostic{
 		Pos: c.pos,
 		End: c.end,
 		Message: fmt.Sprintf(
-			"const %q used as a key at http.Header, but %q is not canonical, want %q",
+			"const %q used as a key at http.Header, but %q is %s, want %q",
```
Comment (ccoVeille) [inline constant_string.go:40]: When I read this `but is %s`

I think that `but is exclusion` might be reported

See https://github.com/lasiar/canonicalheader/pull/60#discussion_r1835526786

So here, maybe `but is excluded` might be better, no?

**Review by ccoVeille — APPROVED** [review]
One suggestion, otherwise 👍

File: analyzer.go [hunk 5]
```diff
@@ -19,19 +20,87 @@ const (
...
 }
 
-func run(pass *analysis.Pass) (any, error) {
+type wellKnownHeader struct {
+	header string
+	reason reasonReport
+}
+
+func (c *canonicalHeader) buildExclusions() map[string]wellKnownHeader {
+	result := make(map[string]wellKnownHeader)
```
Comment (psyhatter) [inline analyzer.go:82]: Could be preallocated

File: configure.go [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+}
+
+func (s *stringSet) Set(flag string) error {
+	list := strings.Split(flag, ",")
+
+	*s = make(stringSet, 0, len(list))
+
+	for _, element := range list {
+		if element == "" {
+			return errors.New("empty string")
```
Comment (psyhatter) [inline configure.go:22]: Can't we just skip such an element?

Comment (psyhatter) [reply configure.go:22]: And it might be useful to `strings.TrimSpace` elements

File: analyzer.go (hunk 3 above)
Comment (lasiar) [reply analyzer.go:61]: I'll simplify the report text.
The reason type will be deleted.

File: cmd/canonicalheader/main.go [hunk 7]
```diff
@@ -7,5 +7,5 @@ import (
 )
 
 func main() {
-	singlechecker.Main(canonicalheader.Analyzer)
```
[the remark this answers is not in the document]
Comment (lasiar) [reply cmd/canonicalheader/main.go:10]: done

File: cmd/initialismer/main.go [hunk 8]
```diff
@@ -367,26 +367,26 @@ func initialism() map[string]string {
 }
 `
 
-const tmplTest = `// Code generated by initialismer; DO NOT EDIT.
+var tmplTest = fmt.Sprintf(`// Code generated by initialismer; DO NOT EDIT.
```
[the remark this answers is not in the document]
Comment (lasiar) [reply cmd/initialismer/main.go:371]: I'll fix it

File: constant_string.go (hunk 4 above)
Comment (lasiar) [reply constant_string.go:40]: https://github.com/lasiar/canonicalheader/pull/60#discussion_r1836657593

File: analyzer.go (hunk 5 above)
Comment (lasiar) [reply analyzer.go:82]: Done

File: configure.go (hunk 6 above)
Comment (lasiar) [reply configure.go:22]: Done!

**Review by psyhatter — APPROVED** [review]
LGTM

File: testdata/src/alias/alias.go.golden [hunk 9]
```diff
@@ -5,5 +5,5 @@ import "net/http"
 type myHeader = http.Header
 
 func _() {
-	myHeader{}.Get("Tt") // want `non-canonical header "TT", instead use: "Tt"`
+	myHeader{}.Get("Tt") // want `use "Tt" instead of "TT"`
```
Comment (ccoVeille) [inline testdata/src/alias/alias.go.golden:8]: Is it normal that every golden files contain a valid code but  the comment want reports an error

Here Tt us used in the code, but the want seems unrelated

File: testdata/src/common/common.go.golden [hunk 10]
```diff
@@ -4,18 +4,18 @@ import (
 	"net/http"
 )
 
-const testHeader = "testHeader"
+const constTestHeader = "testHeaderValue"
 
 func p() {
 	v := http.Header{}
-	v.Get(testHeader) // want `const "testHeader" used as a key at http.Header, but "testHeader" is not canonical, want "Testheader"`
+	v.Get(constTestHeader) // want `use "Testheadervalue" instead of "testHeaderValue"`
```
Comment (ccoVeille) [inline testdata/src/common/common.go.golden:11]: The previous code was reporting it was a constant, here the error message is strange

File: testdata/src/alias/alias.go.golden (hunk 9 above)
Comment (lasiar) [reply testdata/src/alias/alias.go.golden:8]: It's okay.

Uses only by tests https://pkg.go.dev/golang.org/x/tools/go/analysis/analysistest#RunWithSuggestedFixes

File: testdata/src/common/common.go.golden (hunk 10 above)
Comment (lasiar) [reply testdata/src/common/common.go.golden:11]: I think it's more simple.

There are fewer words here and more robotic.

You can understand what the problem is by taking a quick look

Comment (ccoVeille) [reply testdata/src/common/common.go.golden:11]: I liked the previous way, but why not

Comment (psyhatter) [reply testdata/src/common/common.go.golden:11]: I'd rather keep it as simple as it is rn. Mentioning a constant might make an impression that the linter insist on renaming the constant, instead of changing its value or using some other value

Comment (lasiar) [reply testdata/src/common/common.go.golden:11]: @ccoVeille, can I close this conversation?

**Review by ccoVeille — APPROVED** [review]

**Comment by ccoVeille** [comment]
Could a new tag be created ? so it would be detected by golangci-lint?

Thanks
